File: internal/stmt.go
package internal

type stmt interface {
	accept(stmtVisitor) R
}

type stmtVisitor interface {
	visitExprStmt(stmt *exprStmt) R
	visitTryCatchStmt(stmt *tryCatchStmt) R
	visitClassicForStmt(stmt *classicForStmt) R
	visitEnhancedForStmt(stmt *enhancedForStmt) R
	visitLetStmt(stmt *letStmt) R
	visitBlockStmt(stmt *blockStmt) R
	visitWhileStmt(stmt *whileStmt) R
	visitReturnStmt(stmt *returnStmt) R
	visitBreakStmt(stmt *breakStmt) R
	visitContinueStmt(stmt *continueStmt) R
	visitIfStmt(stmt *ifStmt) R
	visitFnStmt(stmt *fnStmt) R
	visitClassStmt(stmt *classStmt) R
}

type exprStmt struct {
	last *token
	expression expr
}

func (s *exprStmt) accept(visitor stmtVisitor) R {
	return visitor.visitExprStmt(s)
}

type tryCatchStmt struct {
	tryBody stmt
	name *token
	catchBody stmt
}

func (s *tryCatchStmt) accept(visitor stmtVisitor) R {
	return visitor.visitTryCatchStmt(s)
}

type classicForStmt struct {
	keyword *token
	initializer stmt
	condition expr
	increment expr
	body stmt
}

func (s *classicForStmt) accept(visitor stmtVisitor) R {
	return visitor.visitClassicForStmt(s)
}

type enhancedForStmt struct {
	keyword *token
	identifiers []*token
	collection expr
	body stmt
}

func (s *enhancedForStmt) accept(visitor stmtVisitor) R {
	return visitor.visitEnhancedForStmt(s)
}

type letStmt struct {
	name *token
	initializer expr
}

func (s *letStmt) accept(visitor stmtVisitor) R {
	return visitor.visitLetStmt(s)
}

type blockStmt struct {
	stmts []stmt
}

func (s *blockStmt) accept(visitor stmtVisitor) R {
	return visitor.visitBlockStmt(s)
}

type whileStmt struct {
	keyword *token
	condition expr
	body stmt
}

func (s *whileStmt) accept(visitor stmtVisitor) R {
	return visitor.visitWhileStmt(s)
}

type returnStmt struct {
	keyword *token
	value expr
}

func (s *returnStmt) accept(visitor stmtVisitor) R {
	return visitor.visitReturnStmt(s)
}

type breakStmt struct {
	keyword *token
}

func (s *breakStmt) accept(visitor stmtVisitor) R {
	return visitor.visitBreakStmt(s)
}

type continueStmt struct {
	keyword *token
}

func (s *continueStmt) accept(visitor stmtVisitor) R {
	return visitor.visitContinueStmt(s)
}

type ifStmt struct {
	keyword *token
	condition expr
	thenBranch []stmt
	elifs []*struct{condition expr; thenBranch []stmt}
	elseBranch []stmt
}

func (s *ifStmt) accept(visitor stmtVisitor) R {
	return visitor.visitIfStmt(s)
}

type fnStmt struct {
	name *token
	params []*token
	body []stmt
}

func (s *fnStmt) accept(visitor stmtVisitor) R {
	return visitor.visitFnStmt(s)
}

type classStmt struct {
	name *token
	superclass *variableExpr
	methods []*fnStmt
	staticMethods []*fnStmt
}

func (s *classStmt) accept(visitor stmtVisitor) R {
	return visitor.visitClassStmt(s)
}

File: rewrite_in_rust/src/stmt.rs
use crate::expr::*;
use crate::tokens::*;

#[derive(Debug, PartialEq, Clone)]
pub struct LetStmt {
    pub name: TokenData,
    pub initializer: Option<Expr>,
}

#[derive(Debug, PartialEq, Clone)]
pub struct ClassicForStmt {
    pub keyword: TokenData,
    pub initializer: Option<Box<Stmt>>,
    pub condition: Expr,
    pub increment: Expr,
    pub body: Box<Stmt>,
}

#[derive(Debug, PartialEq, Clone)]
pub struct EnhancedForStmt {
    pub keyword: TokenData,
    pub identifiers: Vec<TokenData>,
    pub collection: Expr,
    pub body: Box<Stmt>,
}

#[derive(Debug, PartialEq, Clone)]
pub struct TryCatchStmt {
    pub try_body: Box<Stmt>,
    pub name: TokenData,
    pub catch_body: Box<Stmt>,
}

#[derive(Debug, PartialEq, Clone)]
pub struct WhileStmt {
    pub keyword: TokenData,
    pub condition: Expr,
    pub body: Box<Stmt>,
}

#[derive(Debug, PartialEq, Clone)]
pub struct ReturnStmt {
    pub keyword: TokenData,
    pub value: Option<Expr>,
}

#[derive(Debug, PartialEq, Clone)]
pub struct BreakStmt {
    pub keyword: TokenData,
}

#[derive(Debug, PartialEq, Clone)]
pub struct ContinueStmt {
    pub keyword: TokenData,
}

#[derive(Debug, PartialEq, Clone)]
pub struct ElifBranch {
    pub condition: Expr,
    pub then_branch: Vec<Stmt>,
}

#[derive(Debug, PartialEq, Clone)]
pub struct IfStmt {
    pub keyword: TokenData,
    pub condition: Expr,
    pub then_branch: Vec<Stmt>,
    pub elifs: Vec<ElifBranch>,
    pub else_branch: Vec<Stmt>,
}

#[derive(Debug, PartialEq, Clone)]
pub struct FnStmt {
    pub name: TokenData,
    pub params: Vec<TokenData>,
    pub body: Vec<Stmt>,
}

#[derive(Debug, PartialEq, Clone)]
pub struct ClassStmt {
    pub name: Option<TokenData>,
    pub methods: Vec<FnStmt>,
    pub static_methods: Vec<FnStmt>,
    pub superclass: Option<VarExpr>,
}

#[derive(Debug, PartialEq, Clone)]
pub struct BlockStmt {
    pub stmts: Vec<Stmt>,
}

#[derive(Debug, PartialEq, Clone)]
pub struct ExprStmt {
    pub last: Option<TokenData>,
    pub expression: Expr,
}

#[derive(Debug, PartialEq, Clone)]
pub enum Stmt {
    Fn(FnStmt),
    Let(LetStmt),
    Block(BlockStmt),
    Class(ClassStmt),
    ClassicFor(ClassicForStmt),
    EnhancedFor(EnhancedForStmt),
    While(WhileStmt),
    If(IfStmt),
    Continue(ContinueStmt),
    Return(ReturnStmt),
    Break(BreakStmt),
    TryCatch(TryCatchStmt),
    Expr(ExprStmt),
}
